repos/system_upgrade/el7toel8/files/rhel_upgrade.py

```diff
@@ -53,28 +63,28 @@ def configure(self):
                 repo.enable()
 
     def run(self):
-        self.base.add_remote_rpms(self.plugin_data['pkgs_info']['local_rpms'])
 
-        for pkg_spec in self.plugin_data['pkgs_info']['to_remove']:
-            try:
-                self.base.remove(pkg_spec)
-            except dnf.exceptions.MarkingError:
-                self.pkgs_notfound.append(pkg_spec)
+        to_install_local = self.plugin_data['pkgs_info']['local_rpms']
+        to_install = self.plugin_data['pkgs_info']['to_install']
+        to_remove = self.plugin_data['pkgs_info']['to_remove']
+        to_upgrade = self.plugin_data['pkgs_info']['to_upgrade']
 
-        for pkg_spec in self.plugin_data['pkgs_info']['to_install']:
-            try:
-                self.base.install(pkg_spec)
-            except dnf.exceptions.MarkingError:
-                self.pkgs_notfound.append(pkg_spec)
+        # Local (on filesystem) packages to be installed.
+        # add_remote_rpms() accepts list of packages
+
+        self.base.add_remote_rpms(to_install_local)
 
-        q = self.base.sack.query().installed()
-        for pkg in q:
-            if pkg.name not in (self.plugin_data['pkgs_info']['to_install']
-                                + self.plugin_data['pkgs_info']['to_remove']):
-                self.base.upgrade(pkg.name)
+        # Packages to be removed
+        self._process_packages(to_remove, self.base.remove)
+        # Packages to be installed
+        self._process_packages(to_install, self.base.install)
+        # Packages to be upgraded
+        self._process_packages(to_upgrade, self.base.upgrade)
 
         self.base.distro_sync()
 
+        print('Packages not found: {}'.format(self.pkgs_notfound))
```

Comment:
```suggestion
        from dnfpluginsextras import logger
        logger.warning('Packages marked by Leapp for installation/removal/upgrade not found in RHEL 8 repos: {}'.format(self.pkgs_notfound))
```

Reply:
What does "not found" actually mean? Not found in RHEL 8 repos? Not found in rpm database of installed packages? I would add this info too to this message.
Instead of writing directly to stdout (or stderr - as this seems more appropriate there), I'd use dnf instance of python logger: https://dnf.readthedocs.io/en/latest/api_common.html.
Example: https://github.com/rpm-software-management/dnf-plugins-extras/blob/master/plugins/system_upgrade.py#L124

Reply:
@bocekm are you sure "dnfpluginsextras" logger is available in RHEL7?

Reply:
`logger.warning('Packages marked by Leapp for installation/removal/upgrade not found in RHEL 8 repos: {}'.format(self.pkgs_notfound))`

Using string formatting in logger is not recommended:

https://reinout.vanrees.org/weblog/2015/06/05/logging-formatting.html

I agree with "stderr" though...

Reply:
@dmnks, @dmach, could you please give a hint how to do logging in a dnf plugin on RHEL 7? And is it the same on RHEL 8? We do not want to use plain print as an output from the plugin for the user.

Reply:
I'm not very familiar with this, but AFAIK, @jrohel has been working on the logging functionality recently.

Reply:
Pinging @jrohel.

Reply:
I eventually realized it is probably not right to continue if something was not found so I am just raising exception with all packages which are not found.

@dmnks does that look ok?

Reply:
@Rezney probably :) @j-mracek or @jrohel, what do you think? (you have to locate the relevant hunk in the diff)